web/lib/SuffixTree.ts

```diff
@@ -0,0 +1,55 @@
+export class SuffixTree {
+    root: Record<string, any>;
+    steps: Array<{ char: string; path: string[] }>;
+    nodeToParent: WeakMap<Record<string, any>, Record<string, any>>;
+
+    constructor() {
+        this.root = {};
+        this.steps = [];
+        this.nodeToParent = new WeakMap();
+        this.nodeToParent.set(this.root, null);
+    }
+
+    addSuffix(suffix: string) {
+        let node = this.root;
+        for (const char of suffix) {
+            if (!node[char]) {
+                node[char] = {};
+                this.steps.push({ char, path: this.getPath(node) });
+                this.nodeToParent.set(node[char], node);
+            }
+            node = node[char];
+        }
+        node.isEnd = true;
+    }
+
+    addString(str: string) {
+        console.log(`Adding string: ${str}`);
+        for (let i = 0; i < str.length; i++) {
+            this.addSuffix(str.slice(i));
+            console.log(`Added suffix: ${str.slice(i)}`);
+        }
+    }
+
+    getPath(node: Record<string, any>): string[] {
+        const path: string[] = [];
+        while (node && node !== this.root) {
+            const parent = this.nodeToParent.get(node);
+            if (parent) {
+                const entry = Object.entries(parent).find(([key, value]) => value === node);
```

Comment:
`entry` seems to have a meaning like `isValueFound`. It also appears we are looking for the first occurrence of a node which has a value equal to itself, if any. Are we guaranteed to have such a node, and what is the significance of it?

Why do we unshift the first element (entry[0]) if such a node is found?